src/libraries/Microsoft.Bcl.TimeProvider/src/System/Threading/Tasks/TimeProviderTaskExtensions.cs

```diff
@@ -0,0 +1,222 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Threading.Tasks
+{
+    /// <summary>
+    /// Provide extensions methods for <see cref="Task"/> operations with <see cref="TimeProvider"/>.
+    /// </summary>
+    public static class TimeProviderTaskExtensions
+    {
+#if !NET8_0_OR_GREATER
+        private sealed class DelayState : TaskCompletionSource<bool>
+        {
+            public DelayState() : base(TaskCreationOptions.RunContinuationsAsynchronously) {}
+            public ITimer Timer { get; set; }
+            public CancellationTokenRegistration Registration { get; set; }
+        }
+
+        private sealed class WaitAsyncState : TaskCompletionSource<bool>
+        {
+            public WaitAsyncState() : base(TaskCreationOptions.RunContinuationsAsynchronously) { }
+            public readonly CancellationTokenSource ContinuationCancellation = new CancellationTokenSource();
+            public CancellationTokenRegistration Registration;
+            public ITimer? Timer;
+        }
+#endif // !NET8_0_OR_GREATER
+
+        /// <summary>Creates a task that completes after a specified time interval.</summary>
+        /// <param name="timeProvider">The <see cref="TimeProvider"/> with which to interpret <paramref name="delay"/>.</param>
+        /// <param name="delay">The <see cref="TimeSpan"/> to wait before completing the returned task, or <see cref="Timeout.InfiniteTimeSpan"/> to wait indefinitely.</param>
+        /// <param name="cancellationToken">A cancellation token to observe while waiting for the task to complete.</param>
+        /// <returns>A task that represents the time delay.</returns>
+        /// <exception cref="System.ArgumentNullException">The <paramref name="timeProvider"/> argument is null.</exception>
+        /// <exception cref="System.ArgumentOutOfRangeException"><paramref name="delay"/> represents a negative time interval other than <see cref="Timeout.InfiniteTimeSpan"/>.</exception>
+        public static Task Delay(this TimeProvider timeProvider, TimeSpan delay, CancellationToken cancellationToken = default)
+        {
+#if NET8_0_OR_GREATER
+            return Task.Delay(delay, timeProvider, cancellationToken);
+#else
+            if (timeProvider == TimeProvider.System)
+            {
+                return Task.Delay(delay, cancellationToken);
+            }
+
+            if (timeProvider is null)
+            {
+                throw new ArgumentNullException(nameof(timeProvider));
+            }
+
+            if (delay != Timeout.InfiniteTimeSpan && delay < TimeSpan.Zero)
+            {
+                throw new ArgumentOutOfRangeException(nameof(delay));
+            }
+
+            if (delay == TimeSpan.Zero)
+            {
+                return Task.CompletedTask;
+            }
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled(cancellationToken);
```

Comment:
for some reason can't get rid of the feeling that either Canceled or Cancellation is misspelled 😄

Reply:
We've standardized on everything having one L except for cancellation with two, so "cancel", "canceled", "canceling", "canceler", "cancellation".